sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DistributionAndOrderingUtils.scala

```diff
@@ -53,16 +63,27 @@ object DistributionAndOrderingUtils {
         query
       }
 
-      val ordering = toCatalystOrdering(write.requiredOrdering, query)
+      val ordering = toCatalystOrdering(write.requiredOrdering, query, funCatalogOpt)
       val queryWithDistributionAndOrdering = if (ordering.nonEmpty) {
-        Sort(ordering, global = false, queryWithDistribution)
+        Sort(
+          ordering.map(ur => resolveTransformExpression(ur).asInstanceOf[SortOrder]),
+          global = false,
+          queryWithDistribution)
       } else {
         queryWithDistribution
       }
 
-      queryWithDistributionAndOrdering
+      // Apply typeCoercionRules
+      SimpleAnalyzer.execute(queryWithDistributionAndOrdering)
 
     case _ =>
       query
   }
+
+  def resolveTransformExpression(expr: Expression): Expression = expr.transform {
+    case TransformExpression(scalarFunc: ScalarFunction[_], arguments, Some(numBuckets)) =>
```

Comment:
I can't recall the details now. When do we need to translate v2 trasnform to `TransformExpression`?

Reply:
`TransformExpression` is introduced in SPARK-37377(#35657), to check input partition compatibility in `V2ScanPartitioningAndOrdering`

Reply:
SPARK-37377(#35657) changes the common code used by V2 reading & writing, converting the `Transform` to the `TransformExpression`. 
If wanted, we can skip `TransformExpression` and resolve `Transform` to `ApplyFunctionExpression` / `Invoke` / `StaticInvoke` directly, but this may cause V2 reading & writing to share less code.